### PR TITLE
[red-knot] rename {Class,Module,Function} => {Class,Module,Function}Literal

### LINKED ISSUE #13863: [red-knot] rename `Type::Class` and `Type::Function` to `Type::ClassLiteral` and `Type::FunctionLiteral`
We've discussed and agreed this rename would be an improvement. It clarifies better what these types represent, and avoids the likely confusion of assuming `Type::Function` is something like `typing.Callable` and `Type::Class` is the representation of a type expression like `str`.

It also avoids the current false parallel between `Type::Class` and `Type::Instance`, where `Type::Class` is a singleton type representing exactly one class definition, but `Type::Instance` includes instances of subclasses.

### PR DESCRIPTION
**Pull request by sharkdp**
## Summary

* Rename `Type::Class` => `Type::ClassLiteral`
* Rename `Type::Function` => `Type::FunctionLiteral`
* Do not rename `Type::Module`
* Remove `*Literal` suffixes in `display::LiteralTypeKind` variants, as per clippy suggestion
* Get rid of `Type::is_class()` in favor of `is_subtype_of(…, 'type')`; modifiy `is_subtype_of` to support this.
* Add new `Type::is_xyz()` methods and use them instead of matching on `Type` variants.

closes #13863 

## Test Plan

New `is_subtype_of_class_literals` unit test.

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types/display.rs
Comment (sharkdp) [inline crates/red_knot_python_semantic/src/types/display.rs:188]: Instead of adding suffixes to the `Class` and `Function` variants, I removed the suffix from the other remaining variants.

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:260]: I wonder if it also makes sense to rename this to `ModuleLiteral`, to distinguish it from `Type::Instance(types.ModuleType)`

File: crates/red_knot_python_semantic/src/types/display.rs
Comment (MichaReiser) [reply crates/red_knot_python_semantic/src/types/display.rs:188]: Makes sense. I think clippy would also start yelling at you if all have the same suffix :)

Comment (sharkdp) [reply crates/red_knot_python_semantic/src/types/display.rs:188]: Exactly (see PR description)

File: crates/red_knot_python_semantic/src/types.rs
Comment (sharkdp) [reply crates/red_knot_python_semantic/src/types.rs:260]: Same thought here. Then I saw that `Module` was not part of `LiteralTypeKind`, so it's not treated as a literal in the `display` mod.

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:260]: I think how we display the type should probably be secondary to how we use and think about the type when it comes to naming. And there are already some confusing things about the `LiteralTypeKind` enum, e.g. that it (correctly) does not contain a variant for `LiteralString`.

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:260]: In #13257 I propose renaming `LiteralTypeKind` to `CondensedDisplayTypeKind` — it might be worth you doing that here or me pulling it out into a standalone PR. I think it's a clearer name

Comment (sharkdp) [reply crates/red_knot_python_semantic/src/types.rs:260]: > renaming `LiteralTypeKind` to `CondensedDisplayTypeKind`

I can do that.

And what do we do with `Type::Module`?

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:260]: > And what do we do with `Type::Module`?

I would vote to rename it to `ModuleLiteral` but otherwise leave it unchanged in this PR. I think if we rename `LiteralTypeKind` to `CondensedDisplayTypeKind`, it becomes more obvious that not all "literal" types really need to be included in that enum, just the ones that have condensed displays

Comment (sharkdp) [reply crates/red_knot_python_semantic/src/types.rs]: Done.

File: crates/red_knot_python_semantic/src/semantic_model.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/semantic_model.rs:208]: nit: we could add `is_function_literal()` and `is_class_literal` methods to `Type`, and use them here

File: crates/red_knot_python_semantic/src/types/display.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/display.rs:182]:
```suggestion
/// Enumeration of literal types that are displayed in a "condensed way" inside `Literal` slices.
///
/// For example, `Literal[1] | Literal[2]` is displayed as `"Literal[1, 2]"`.
/// Not all `Literal` types are displayed using `Literal` slices
/// (e.g. it would be inappropriate to display `LiteralString`
/// as `Literal[LiteralString]`).
#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
```

**Review by AlexWaygood — APPROVED** [review]

**Comment by AlexWaygood** [comment]
(Could possibly wait for @carljm's review as well on this before landing)

File: crates/red_knot_python_semantic/src/semantic_model.rs
Comment (sharkdp) [reply crates/red_knot_python_semantic/src/semantic_model.rs:208]: This method does not exist :smile:. I could do `assert!(ty.into_function_literal_type().is_some())`, but I don't think that's much better than matching on the variant?

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/semantic_model.rs:208]: > This method does not exist 😄

I know! Sorry for being unclear. I'm suggesting we could add `Type::is_function_literal()` and `Type::is_class_literal()` methods to the `Type` enum, for consistency with our existing `Type::is_unbound` and `Type::is_never` methods. It doesn't _have_ to be done in this PR, but it also _might as well_ if we're touching these lines anyway ;)

Comment (sharkdp) [reply crates/red_knot_python_semantic/src/semantic_model.rs:208]: No, sorry. I misread.

`is_function_literal` does not exist, but `is_class_literal` exists. It does something different from just matching against `Type::ClassLiteral` though :see_no_evil: 

```rs
pub fn is_class_literal(&self, db: &'db dyn Db) -> bool {
    match self {
        Type::Union(union) => union.elements(db).iter().all(|ty| ty.is_class_literal(db)),
        Type::ClassLiteral(_) => true,
        // / TODO include type[X], once we add that type
        _ => false,
    }
}
```

So if we want to do this properly, I guess we would have to rename the existing `is_class_literal` function...

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/semantic_model.rs:208]: > So if we want to do this properly, I guess we would have to rename the existing `is_class_literal` function...

We're well down the renaming rabbit hole now. May as well 😆

It seems like the question the existing `is_class_literal` method is _actually_ trying to answer is "Is the type a subtype of `Instance(builtins.type)`?". Maybe we could try to just get rid of it and make `is_subtype_of` work correctly for `Instance(builtins.type)`, so that it recognises all `Type::ClassLiteral`s as subtypes

**Review by carljm — APPROVED** [review]
Looks great to me, thank you!

File: crates/red_knot_python_semantic/src/semantic_model.rs
Comment (sharkdp) [reply crates/red_knot_python_semantic/src/semantic_model.rs]: Did that now.

**Review by carljm — APPROVED** [review]
Looks excellent!

**Pull request merged**
